pycbc/vetoes/chisq_cupy.py

```diff
@@ -0,0 +1,341 @@
+
+    // load integration boundaries (might not be bin boundaries if bin is large)
+    if (threadIdx.x == 0){
+        s = kmin[blockIdx.x];
+        e = kmax[blockIdx.x];
+    }
+
+    % for p in range(NP):
+        chisq[threadIdx.x + ${NT*p}].x = 0;
+        chisq[threadIdx.x + ${NT*p}].y = 0;
+    % endfor
+    __syncthreads();
+
+    // calculate the chisq integral for each thread
+    // sliding reduction for each thread from s, e
+    for (int i = threadIdx.x + s; i < e; i += blockDim.x){
+        float re, im;
+
+        %if fuse:
+            float2 qt, st, ht;
+            st = stilde[i];
+            ht = htilde[i];
+            qt.x = ht.x * st.x + ht.y * st.y;
+            qt.y = ht.x * st.y - ht.y * st.x;
+        %else:
+            float2 qt = corr[i];
+        %endif
+
+        %for p in range(NP):
+            sincosf(phase${p} * i, &im, &re);
+            chisq[threadIdx.x + ${NT*p}].x += re * qt.x - im * qt.y;
+            chisq[threadIdx.x + ${NT*p}].y += im * qt.x + re * qt.y;
+        %endfor
+    }
+
+    float x, y, x2, y2;
+    // logarithmic reduction within thread block
+    for (int j=${NT} / 2; j>=1; j/=2){
+        if (threadIdx.x <j){
+            %for p in range(NP):
+                __syncthreads();
+                x = chisq[threadIdx.x + ${NT*p}].x;
+                y = chisq[threadIdx.x + ${NT*p}].y;
+                x2 = chisq[threadIdx.x + j + ${NT*p}].x;
+                y2 = chisq[threadIdx.x + j + ${NT*p}].y;
+                 __syncthreads();
+                chisq[threadIdx.x + ${NT*p}].x = x + x2;
+                chisq[threadIdx.x + ${NT*p}].y = y + y2;
+            %endfor
+        }
+    }
+
+    if (threadIdx.x == 0){
+        % for p in range(NP):
+            atomicAdd(&outc[bv[blockIdx.x] + nbins * ${p}].x, chisq[0 + ${NT*p}].x);
+            atomicAdd(&outc[bv[blockIdx.x] + nbins * ${p}].y, chisq[0 + ${NT*p}].y);
+        % endfor
+    }
+
+}
+""")
+
+chisqkernel_pow2 = Template("""
+#include <cstdint>
+extern "C" __global__ void power_chisq_at_points_${NP}_pow2(
+                                      %if fuse:
+                                          float2* htilde,
+                                          float2* stilde,
+                                      %else:
+                                          float2* corr,
+                                      %endif
+                                      float2* outc, unsigned int N,
+                                      %for p in range(NP):
+                                        unsigned int points${p},
+                                      %endfor
+                                      uint32_t* kmin,
+                                      uint32_t* kmax,
+                                      uint32_t* bv,
+                                      unsigned int nbins){
+    __shared__ unsigned int s;
+    __shared__ unsigned int e;
+    __shared__ float2 chisq[${NT} * ${NP}];
+    float twopi = 6.283185307179586f;
```

Comment:
I don't really like that this is defined here - is there any SSOT we can get this from automatically?

Reply:
I can pull it from LAL and then update the source code when run ...